omod/src/main/java/org/openmrs/module/pihcore/fragment/controller/dashboardwidgets/DiscreteValuesFragmentController.java

```diff
@@ -0,0 +1,143 @@
+
+    public void controller(
+                           @SpringBean("conceptService") ConceptService conceptService,
+                           @SpringBean("obsService") ObsService obsService,
+                           @SpringBean("programWorkflowService") ProgramWorkflowService programWorkflowService,
+                           @FragmentParam("app") AppDescriptor app,
+                           FragmentConfiguration config,
+                           FragmentModel model) {
+
+        Object patientConfig = config.get("patient");
+        Patient patient = null;
+        if (patientConfig != null && (patientConfig instanceof PatientDomainWrapper)) {
+            patient = ((PatientDomainWrapper) patientConfig).getPatient();
+        }
+        if (patient == null) {
+            throw new IllegalArgumentException("No patient found. Please pass patient into the configuration");
+        }
+
+        model.put("app", app);
+        model.put("patient", patient);
+
+        JsonNode node = app.getConfig().get("fields");
+        if (node == null) {
+            throw new IllegalStateException("Missing configuration concepts on widget");
+        }
+        Set<Concept> concepts = new HashSet<>();
+        Map<String, SimpleObject> fields = new LinkedHashMap<>();
+        if (node instanceof ArrayNode) {
+            ArrayNode arrayNode = (ArrayNode) node;
+            for (Iterator<JsonNode> iter = arrayNode.getElements(); iter.hasNext();) {
+                JsonNode fieldNode = iter.next();
+                try {
+                    String label = fieldNode.get("label").asText();
+                    String conceptUuid = fieldNode.get("concept").asText();
+                    String minValue = fieldNode.get("minValue").asText();
+                    if (StringUtils.isNotBlank(conceptUuid)) {
+                        Concept concept = conceptService.getConceptByUuid(conceptUuid);
+                        if (concept == null ) {
+                            throw new IllegalArgumentException("No concept with this UUID: " + conceptUuid + " found. Please pass correct concept UUID into the configuration");
+                        }
+                        concepts.add(concept);
+                        fields.put(conceptUuid, SimpleObject.create(
+                                "label", label,
+                                "minValue", minValue
+                        ));
+                    }
+                }
+                catch (Exception e) {
+                    throw new IllegalStateException("Invalid configuration for encounterType node", e);
+                }
+            }
+        } else {
+            throw new IllegalStateException("fields is expected to be an array of objects");
+        }
+
+        Date obsOnOrAfter = null;
+        String visitUrl = getConfigValue(app, "visitUrl");
+        String programUuid = getConfigValue(app, "duringCurrentEnrollmentInProgram");
```

Comment:
At this point, it is probably high time that we move some of what we are doing here into a common set of utility methods.  We now have a number of controllers that all have similar logic for getting encounter and obs that fall within the current enrollment of a particular program.  And we will likely continue to need more of this.  Can we look at extracting this to some utilities that we can re-use, test, and ensure bug fixes are applied universally?